src/components/DotNavigation/DotNavigation.tsx

```diff
@@ -0,0 +1,280 @@
+      length,
+    )
+      ? {
+          ...prevState,
+          savedCurrentIndex: currentIndex,
+          ...(DotNavigation.shouldAnimate(
+            currentIndex,
+            savedCurrentIndex,
+            length,
+          ) && {
+            animation:
+              currentIndex < savedCurrentIndex
+                ? Animation.Back
+                : Animation.Forward,
+            longListKey: getUniqueLongListKey(longListKey),
+          }),
+        }
+      : prevState;
+  };
+
+  static getDerivedStateFromProps = (
+    nextProps: DotNavigationProps,
+    prevState: DotNavigationState,
+  ) =>
+    nextProps.length <= MAX_SHORT_LIST_LENGTH
+      ? prevState
+      : DotNavigation.getLongListAnimation(nextProps, prevState);
+
+  callOnSelect = (index: number) => () => this.props.onSelect(index);
+
+  getRadio = ({ key, checked, size, fakeRadio }: RadioProps) => (
+    <div className={classNames(styles.radio)}>
+      <div
+        key={key}
+        className={classNames(styles.dot, {
+          [styles.small]: size === Size.Small,
+          [styles.tiny]: size === Size.Tiny,
+          [styles.bordered]: this.props.showBorder === true,
+          [styles.light]: this.props.theme === Theme.Light,
+          [styles.checked]: checked,
+        })}
+        data-hook={fakeRadio ? undefined : DotNavigationDataHooks.Dot}
+      />
+    </div>
+  );
+
+  renderFakeRadio = (key: string) =>
+    this.getRadio({ key, checked: false, size: Size.Tiny, fakeRadio: true });
+
+  renderRadioButton = ({
+    key,
+    checked,
+    size,
+    onChange,
+    ariaLabel,
+  }: RadioButtonProps) => (
+    <RadioButton
+      key={key}
+      checked={checked}
+      checkedIcon={this.getRadio({ checked, size })}
+      uncheckedIcon={this.getRadio({ checked: false, size })}
+      onChange={onChange}
+      aria-label={ariaLabel}
+      {...{ className: styles.extendedRadioButton }}
+    />
+  );
+
+  renderList = (values: number[], sizes: Size[]) =>
+    values.map((value, index) =>
+      this.renderRadioButton({
+        key: index,
+        checked: value === this.props.currentIndex,
+        size: sizes[index],
+        onChange: this.callOnSelect(value),
+        ariaLabel: `${this.props.ariaLabel} ${value}`,
+      }),
+    );
+
+  renderStartList = () => this.renderList(getFiveFromStart(0), START_SIZES);
+
+  renderEndList = () =>
+    this.renderList(getFiveFromStart(this.props.length - 5), END_SIZES);
+
+  renderMiddleList = () =>
+    this.renderList(
+      getFiveFromStart(this.props.currentIndex - 2),
+      MIDDLE_SIZES,
+    );
+
+  renderLongVersion = () => (
+    <div
+      className={classNames(styles.radioGroup, {
+        [styles.back]: this.state.animation === Animation.Back,
+        [styles.forward]: this.state.animation === Animation.Forward,
+      })}
+      key={this.state.longListKey}
+    >
+      {this.renderFakeRadio('fake-start')}
+      {!isNumber(this.props.currentIndex) || this.props.currentIndex < 2
```

Comment:
And by the way, in general I think it can be something like:
```js
const NUM_OF_NAV_DOTS = 2; // for the small and medium dot (the name should be something better than this)
this.props.currentIndex < MAX_SHORT_LIST_LENGTH - NUM_OF_NAV_DOTS;
...
this.props.currentIndex >= this.props.length - (MAX_SHORT_LIST_LENGTH - NUM_OF_NAV_DOTS);
```

Reply:
Done)